crates/protogen/src/metastore/types/options.rs

```diff
@@ -424,6 +420,33 @@ impl From<DeltaLakeUnityCatalog> for options::DeltaLakeUnityCatalog {
     }
 }
 
+// Options for a generic `ObjectStore`; to make them as versatile and compact as possible it's just
+// a wrapper for a map, like in `delta-rs`, except here it's a `BTreeMap` instead of a `HashMap`,
+// since the former is Hash unlike the latter. This enables us to capture a variety of different
+// parameters across different object stores, as well as support different optional parameters in
+// differing use-cases (e.g. creating a Delta database vs a Delta table).
+// List of supported config options in `object_store` crate by object store type:
+// - [Azure options](https://docs.rs/object_store/latest/object_store/azure/enum.AzureConfigKey.html#variants)
+// - [S3 options](https://docs.rs/object_store/latest/object_store/aws/enum.AmazonS3ConfigKey.html#variants)
+// - [Google options](https://docs.rs/object_store/latest/object_store/gcp/enum.GoogleConfigKey.html#variants)
+#[derive(Debug, Default, Clone, Arbitrary, PartialEq, Eq, Hash)]
+pub struct StorageOptions {
+    pub inner: BTreeMap<String, String>,
+}
```

Comment:
This is the core of the proposed implementation.

For a bit of context, I initially went with hard-coded structs to try and capture the various storage options in various situations. That turned out to be too verbose and repetitive for my liking, with having to cover a lot of different dimensions (external DBs vs tables, Delta vs Iceberg, GCS vs S3 vs ...). 

So I ended up going with this solution instead, which has significantly less complexity I believe. Arguably we lose some amount of transparency and validation at each step in the pipeline, but the counter-argument would be that we should do validation at initial setup anyway (i.e. when creating the external DB/table in the first place).

In addition, there's a great deal of overlap with how we treat regular files (`TableOptions::S3`, `TableOptions::Gcs`) on object stores and the proposed Delta/Iceberg storage options handling. The present implementation could conceivably be used to consolidate those as well (though I didn't want to try that without a discussion, or in this PR for that matter).

Reply:
I agree that this is much simpler than explicitly listing out the options. We originally did that because the first implementation was for GCS, and all that needed was the service account and location, and then we had S3, which now needed more stuff. So really the code we ended up with was an evolution of adding stuff over time and not planning it up front.

I also agree there is a good bit of overlap with table options. I had a TODO or comment (or even slack message) saying that I wanted to try to consolidate the individual object store files with the data lake stuff. I would be in favor of moving towards that.

Reply:
Got it, makes sense.